CMakeLists.txt

```diff
@@ -5,15 +5,87 @@ project("SVF")
-set(CMAKE_CXX_STANDARD 14)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-# add -std=gnu++14
-set(CMAKE_CXX_EXTENSIONS ON)
+# Match configuration used to build LLVM (match C++ standard; match
+# runtime typing information (RTTI); match exception handling; etc)
+if(COMMAND add_llvm_library)
+  message(STATUS "Detected in-tree build configuration; skipping LLVM fetching")
+  set(IN_SOURCE_BUILD 1)
+else()
+  message(STATUS "Detected out-of-tree build configuration; fetching LLVM")
 
-add_compile_options("-fno-rtti")
-add_compile_options("-fno-exceptions")
+  find_package(LLVM REQUIRED CONFIG HINTS ${LLVM_DIR} $ENV{LLVM_DIR})
+  message(STATUS "LLVM STATUS:
+    Version       ${LLVM_VERSION}
+    Definitions   ${LLVM_DEFINITIONS}
+    Includes      ${LLVM_INCLUDE_DIRS}
+    Libraries     ${LLVM_LIBRARY_DIRS}
+    Targets       ${LLVM_TARGETS_TO_BUILD}
+    Build type    ${LLVM_BUILD_TYPE}
+    Exceptions    ${LLVM_ENABLE_EH}
+    RTTI          ${LLVM_ENABLE_RTTI}
+    Dynamic lib   ${LLVM_LINK_LLVM_DYLIB}"
+  )
+
+  if(CMAKE_BUILD_TYPE STREQUAL "Debug" AND NOT ${LLVM_BUILD_TYPE} STREQUAL "Debug")
+    message(NOTICE "Building SVF in debug-mode but LLVM was not built in debug-mode; "
+                   "debug information could be incomplete when using SVF from LLVM")
+  endif()
+
+  if(NOT LLVM_ENABLE_EH)
+    message(STATUS "Building SVF without exception handling")
+    add_compile_options("-fno-exceptions")
+  endif()
+
+  if(NOT LLVM_ENABLE_RTTI)
+    message(STATUS "Building SVF without RTII")
+    add_compile_options("-fno-rtti")
+  endif()
+
+  # Load the LLVM definitions & include directories
+  separate_arguments(LLVM_DEFINITIONS_LIST NATIVE_COMMAND ${LLVM_DEFINITIONS})
+
+  if(NOT ${LLVM_LINK_LLVM_DYLIB})
+    message(STATUS "Linking to separate LLVM static libraries")
+    llvm_map_components_to_libnames(
+      llvm_libs
+      analysis
+      bitwriter
+      core
+      instcombine
+      instrumentation
+      ipo
+      irreader
+      linker
+      scalaropts
+      support
+      target
+      transformutils)
+  else()
+    message(STATUS "Linking to LLVM dynamic shared library object")
+    set(llvm_libs LLVM)
+  endif()
+
+  # Make the "add_llvm_library()" command available
+  include(AddLLVM)
+
+  # Set the default C++ standard used to build SVF; LLVM versions up to version 6 defaulted to C++98,
+  # and up to (and including) version 15 the default was C++14. From versions 16 and onwards, the
+  # default C++ standard has been C++17. Build SVF with C++14 unless the LLVM version is below 6.
+  if (LLVM_VERSION_MAJOR VERSION_LESS 6)
+    set(CMAKE_CXX_STANDARD 98)
+    message(STATUS "Got LLVM version ${LLVM_VERSION}; using C++ standard: ${CMAKE_CXX_STANDARD}")
+  elseif(LLVM_VERSION_MAJOR VERSION_LESS 16)
+    set(CMAKE_CXX_STANDARD 14)
+    message(STATUS "Got LLVM version ${LLVM_VERSION}; using C++ standard: ${CMAKE_CXX_STANDARD}")
+  else()
+    message(FATAL_ERROR "Found unsupported LLVM version; need <= 15, got: ${LLVM_VERSION}")
```

Comment:
We will update to LLVM-16 in the coming weeks. Would be good to have this `else` branch open with C++14 for now?

Reply:
Interesting; I thought upgrading to LLVM-16 or higher was not planned due to the switch to opaque pointers? Either way, I'll modify this check to a warning for versions greater than 16. :+1:

Reply:
Thanks.

Reply:
Done; for versions greater than 15 the C++17 standard will now be used (a warning will be printed for now, feel free to remove that when the switch to version LLVM 16 is released).